File: src/petrify_converter/color_extractor.py
# src/petrify_converter/color_extractor.py
import io
from statistics import median

from PIL import Image


class ColorExtractor:
    """mainBmp 이미지에서 색상 추출."""

    BACKGROUND_COLORS = {"#ffffff", "#fffff0"}

    def __init__(self, image_data: bytes):
        self.image = Image.open(io.BytesIO(image_data)).convert('RGBA')
        self.pixels = self.image.load()
        self.width, self.height = self.image.size

    def get_color_at(self, x: int, y: int) -> tuple[str, int]:
        """좌표에서 색상과 알파값 추출.

        Returns:
            (hex_color, alpha) 튜플
        """
        if not (0 <= x < self.width and 0 <= y < self.height):
            return "#000000", 255

        r, g, b, a = self.pixels[x, y]
        hex_color = f"#{r:02x}{g:02x}{b:02x}"
        return hex_color, a

    def get_width_at(self, x: int, y: int) -> int:
        """포인트에서 스트로크 굵기 측정 (4방향, alpha > 0 기준).

        Returns:
            굵기 (px). 투명이거나 범위 벗어나면 0.
        """
        if not (0 <= x < self.width and 0 <= y < self.height):
            return 0

        if self.pixels[x, y][3] == 0:  # 투명
            return 0

        # 수직 측정
        v_width = 1
        for dy in [-1, 1]:
            cy = y + dy
            while 0 <= cy < self.height and self.pixels[x, cy][3] > 0:
                v_width += 1
                cy += dy

        # 수평 측정
        h_width = 1
        for dx in [-1, 1]:
            cx = x + dx
            while 0 <= cx < self.width and self.pixels[cx, y][3] > 0:
                h_width += 1
                cx += dx

        return min(v_width, h_width)

    def extract_stroke_width(self, points: list[list]) -> int:
        """스트로크 포인트들의 대표 굵기 추출 (이상치 제거 후 중앙값).

        Args:
            points: [[x, y, timestamp], ...] 형식

        Returns:
            굵기 (px). 측정 불가시 기본값 1.
        """
        widths = []
        for point in points:
            x, y = int(point[0]), int(point[1])
            w = self.get_width_at(x, y)
            if w > 0:
                widths.append(w)

        if not widths:
            return 1

        # 이상치 필터링 적용
        filtered = self._filter_outliers(widths)
        if not filtered:
            return 1

        return int(median(filtered))

    def _filter_outliers(self, values: list[int]) -> list[int]:
        """IQR 기반으로 이상치 제거.

        Args:
            values: 측정된 굵기 값들

        Returns:
            이상치가 제거된 값들. 4개 미만이면 그대로 반환.
        """
        if len(values) < 4:
            return values

        sorted_vals = sorted(values)
        q1_idx = len(sorted_vals) // 4
        q3_idx = 3 * len(sorted_vals) // 4
        q1 = sorted_vals[q1_idx]
        q3 = sorted_vals[q3_idx]
        iqr = q3 - q1

        lower = q1 - 1.5 * iqr
        upper = q3 + 1.5 * iqr

        return [v for v in values if lower <= v <= upper]

File: src/color-extractor.ts
export class ColorExtractor {
  static readonly BACKGROUND_COLORS = new Set(['#ffffff', '#fffff0']);
  static readonly OUTLIER_THRESHOLD = 1.5;
  static readonly LOWER_PERCENTILE = 5;

  private readonly width: number;
  private readonly height: number;
  private readonly data: Uint8ClampedArray;

  constructor(imageData: ImageData) {
    this.width = imageData.width;
    this.height = imageData.height;
    this.data = imageData.data;
  }

  static async fromPng(pngData: ArrayBuffer): Promise<ColorExtractor> {
    const blob = new Blob([pngData], { type: 'image/png' });
    const bitmap = await createImageBitmap(blob);
    const canvas = new OffscreenCanvas(bitmap.width, bitmap.height);
    const ctx = canvas.getContext('2d')!;
    ctx.drawImage(bitmap, 0, 0);
    const imageData = ctx.getImageData(0, 0, bitmap.width, bitmap.height);
    return new ColorExtractor(imageData);
  }

  private isOutOfBounds(x: number, y: number): boolean {
    return x < 0 || x >= this.width || y < 0 || y >= this.height;
  }

  private getPixelIndex(x: number, y: number): number {
    return (y * this.width + x) * 4;
  }

  private getAlphaAt(x: number, y: number): number {
    return this.data[this.getPixelIndex(x, y) + 3];
  }

  getColorAt(x: number, y: number): { color: string; opacity: number } {
    if (this.isOutOfBounds(x, y)) {
      return { color: '#000000', opacity: 255 };
    }

    const idx = this.getPixelIndex(x, y);
    const r = this.data[idx];
    const g = this.data[idx + 1];
    const b = this.data[idx + 2];
    const a = this.data[idx + 3];

    return { color: this.rgbToHex(r, g, b), opacity: a };
  }

  private rgbToHex(r: number, g: number, b: number): string {
    return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
  }

  getWidthAt(x: number, y: number): number {
    if (this.isOutOfBounds(x, y) || this.getAlphaAt(x, y) === 0) {
      return 0;
    }

    const vWidth = this.measureWidth(x, y, 0, 1);
    const hWidth = this.measureWidth(x, y, 1, 0);

    return Math.min(vWidth, hWidth);
  }

  private measureWidth(x: number, y: number, dx: number, dy: number): number {
    let width = 1;
    for (const direction of [-1, 1]) {
      let cx = x + dx * direction;
      let cy = y + dy * direction;
      while (!this.isOutOfBounds(cx, cy) && this.getAlphaAt(cx, cy) !== 0) {
        width++;
        cx += dx * direction;
        cy += dy * direction;
      }
    }
    return width;
  }

  extractStrokeWidth(points: number[][]): number {
    const widths: number[] = [];
    for (const point of points) {
      const w = this.getWidthAt(Math.floor(point[0]), Math.floor(point[1]));
      if (w > 0) widths.push(w);
    }

    if (widths.length === 0) return 1;

    const filtered = this.filterOutliers(widths.sort((a, b) => a - b));
    const idx = Math.floor(filtered.length / ColorExtractor.LOWER_PERCENTILE);
    return filtered[idx];
  }

  private filterOutliers(sortedWidths: number[]): number[] {
    if (sortedWidths.length <= 1) return sortedWidths;

    const filtered = [sortedWidths[0]];
    for (let i = 1; i < sortedWidths.length; i++) {
      if (sortedWidths[i] > sortedWidths[i - 1] * ColorExtractor.OUTLIER_THRESHOLD) {
        break;
      }
      filtered.push(sortedWidths[i]);
    }
    return filtered;
  }
}
